File: ios/Runner/api/MethodChannelResponse.swift
//
//  MethodChannelResponse.swift
//  Runner
//
//  Created by SamLS Chen on 2024/8/20.
//

import Foundation


class MethodChannelResponse<T> {
    
    let statusCode: Int
    let errorMessage: String?
    let data: T?
    
    init (serviceResult: ServiceResult<T>) {
        guard serviceResult.error == nil else {
            switch(serviceResult.error!) {
            case .UnsupportedMediaTypeError:
                statusCode = 10000
            case .BadRequestError:
                statusCode = 10001
            case .UnknownError(_):
                statusCode = 10003
            }
            data = nil
            errorMessage = serviceResult.error!.message
            return
        }
        
        statusCode = 0
        data = serviceResult.data
        errorMessage = ""
    }
}

File: plugin/download/android/src/main/kotlin/com/example/perfect_corp_homework/download/api/MethodChannelResponse.kt
package com.example.perfect_corp_homework.download.api

import com.example.perfect_corp_homework.download.exception.BadRequestError
import com.example.perfect_corp_homework.download.exception.JsonSerializationError
import com.example.perfect_corp_homework.download.exception.NoInternetError
import com.example.perfect_corp_homework.download.exception.UnSupportImageTypeError
import com.example.perfect_corp_homework.download.exception.UnknownError

class MethodChannelResponse<T>(serviceResult: ServiceResult<T>) {

    val statusCode: Int = when(serviceResult.error) {
        is UnSupportImageTypeError -> 10000
        is BadRequestError -> 10001
        is NoInternetError -> 10002
        is UnknownError -> 10003
        is JsonSerializationError -> 10004
        is Exception -> 10003
        else -> 0
    }
    val errorMessage: String? = serviceResult.error?.message
    val data: T? = serviceResult.data
}